akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/PersistentBehavior2.scala

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.typed.javadsl
+
+import java.util.function.Predicate
+import java.util.{ Collections, Optional }
+
+import akka.actor.typed
+import akka.actor.typed.{ BackoffSupervisorStrategy, Behavior }
+import akka.actor.typed.Behavior.DeferredBehavior
+import akka.annotation.{ ApiMayChange, InternalApi }
+import akka.persistence.SnapshotMetadata
+import akka.persistence.typed.{ EventAdapter, _ }
+import akka.persistence.typed.internal._
+import scala.util.{ Failure, Success }
+
+/** Java API */
+@ApiMayChange
+abstract class PersistentBehavior2[Command, Event, State >: Null] private (val persistenceId: String, supervisorStrategy: Option[BackoffSupervisorStrategy]) extends DeferredBehavior[Command] {
+
+  def this(persistenceId: String) = {
+    this(persistenceId, None)
+  }
+
+  def this(persistenceId: String, backoffSupervisorStrategy: BackoffSupervisorStrategy) = {
+    this(persistenceId, Some(backoffSupervisorStrategy))
+  }
+
+  /**
+   * Factory of effects.
+   *
+   * Return effects from your handlers in order to instruct persistence on how to act on the incoming message (i.e. persist events).
+   */
+  protected final def Effect: EffectFactories[Command, Event, State] = EffectFactory.asInstanceOf[EffectFactories[Command, Event, State]]
+
+  /**
+   * Implement by returning the initial empty state object.
+   * This object will be passed into this behaviors handlers, until a new state replaces it.
+   *
+   * Also known as "zero state" or "neutral state".
+   */
+  protected def emptyState: State
+
+  /**
+   * Implement by handling incoming commands and return an `Effect()` to persist or signal other effects
+   * of the command handling such as stopping the behavior or others.
+   *
+   * The command handlers are only invoked when the actor is running (i.e. not replaying).
+   * While the actor is persisting events, the incoming messages are stashed and only
+   * delivered to the handler once persisting them has completed.
+   */
+  protected final def commandHandler(): CommandHandler[Command, Event, State] = {
+    val builder = new HandlerBuilders.CommandHandlerBuilder[Command, Event, State]()
+    commandHandlerSetup(builder)
+    builder.build()
+  }
+
+  protected def commandHandlerSetup(builder: HandlerBuilders.CommandHandlerBuilder[Command, Event, State]): Unit
```

Comment:
Novelty here is that the user need to implement the setup. We build a builder and pass it here. 

The user only needs to add the cases. 

We take over the job of creating the builder and building it.

Reply:
yes, that could be the way to go, although it enforces the use of builders while we said that there could be some value in making that optional (we could have that as a lower level `PersistentBehaviorBase`)

Reply:
True, I didn't think about that.